### PR TITLE
Format range

### PR DESCRIPTION
**Pull request by akshita31**
Issue : https://github.com/OmniSharp/omnisharp-vscode/issues/214

On selecting after the first (alphanumeric) character in "Format Selection" , the code was not trimming the preceding white spaces in the first line of the selection. Modified the code to set the start of the span to the full span of the intersecting token and added the relevant test cases.

Please review : @DustinCampbell @TheRealPiotrP @rchande

### CODE REVIEW

File: src/OmniSharp.Roslyn.CSharp/Services/Formatting/FormatRangeService.cs [hunk 1]
```diff
@@ -26,11 +26,12 @@ public FormatRangeService(OmniSharpWorkspace workspace)
             {
                 return null;
             }
-
+           
             var text = await document.GetTextAsync();
             var start = text.Lines.GetPosition(new LinePosition(request.Line, request.Column));
             var end = text.Lines.GetPosition(new LinePosition(request.EndLine, request.EndColumn));
-            var changes = await FormattingWorker.GetFormattingChanges(document, start, end);
+            var tokenStart = document.GetSyntaxRootAsync().Result.FindToken(start).FullSpan.Start;
```
Comment (rchande) [inline src/OmniSharp.Roslyn.CSharp/Services/Formatting/FormatRangeService.cs]: await

File: src/OmniSharp.Roslyn.CSharp/Services/Formatting/FormatRangeService.cs [hunk 2]
```diff
@@ -26,11 +26,12 @@ public FormatRangeService(OmniSharpWorkspace workspace)
             {
                 return null;
             }
-
+
```
Comment (rchande) [inline src/OmniSharp.Roslyn.CSharp/Services/Formatting/FormatRangeService.cs]: Remove spaces

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs [hunk 3]
```diff
@@ -128,6 +128,85 @@ public void M ()
                 new LinePositionSpanTextChange { StartLine = 2, StartColumn = 30, EndLine = 3, EndColumn = 0, NewText = "\n " });
         }
 
+
```
Comment (rchande) [inline tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: Remove extra blanks

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs [hunk 4]
```diff
@@ -128,6 +128,85 @@ public void M ()
...
+                "{",
+                "    public static void Main()",
+                "    {",
+                "[|               int foo = 1;|]",
+                "    }",
+                "}",
+            };
+
+            await AssertTextChanges(string.Join("\r\n", source),
+                new LinePositionSpanTextChange { StartLine = 3, StartColumn = 5, EndLine = 4, EndColumn = 7, NewText = "\n" });
```
Comment (rchande) [inline tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: This is really hard to read. How hard would it be to make it so the text change got applied and you could just specify the actual expected text?

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs [hunk 5]
```diff
@@ -128,6 +130,116 @@ public void M ()
...
+    }
+}";
+
+            await AssertTextChanges(string.Join("\r\n", source), expected);
+        }
+
+        [Fact]
+        public async Task TextChangesOnStartingSpanAfterFirstCharacterInLine()
+        {
+            var source = new[]
```
Comment (rchande) [inline tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: I think you are storing all of these as string arrays and immediately string.join ing them. Might as well use string literals...

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs [hunk 6]
```diff
@@ -200,5 +312,51 @@ private static void AssertFormatTargetKind(SyntaxKind kind, string input)
...
+                var textChanges = GetTextChanges(oldText, response.Changes);
+                var actualText = oldText.WithChanges(textChanges).ToString();
+                Assert.Equal(expected.Replace("\r\n","\n"), actualText.Replace("\r\n","\n"));
+            }
+        }
+
+        public static IEnumerable<TextChange> GetTextChanges(SourceText oldText, IEnumerable<LinePositionSpanTextChange> changes)
+        {
+            var textChanges = new List<TextChange>();
+            foreach( var change in changes)
```
Comment (rchande) [inline tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: extra space

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs [hunk 7]
```diff
@@ -200,5 +312,51 @@ private static void AssertFormatTargetKind(SyntaxKind kind, string input)
...
+            {
+                var startPosition = new LinePosition(change.StartLine, change.StartColumn);
+                var endPosition = new LinePosition(change.EndLine, change.EndColumn);
+                var span = oldText.Lines.GetTextSpan(new LinePositionSpan(startPosition, endPosition));
+                var newText = change.NewText;
+                textChanges.Add(new TextChange(span, newText));
+            }
+
+            return textChanges.OrderBy(change => change.Span);
```
Comment (rchande) [inline tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: I don't think this is necessary.

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs [hunk 8]
```diff
@@ -128,6 +130,116 @@ public void M ()
...
+    {
+        int foo = 1;
+        bool b = false;
+        Console.WriteLine(foo);
+    }
+}";
+
+            await AssertTextChanges(string.Join("\r\n", source), expected);
+        }
+
```
Comment (rchande) [inline tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: Extra blanks

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs (hunk 7 above)
Comment (akshita31) [reply tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: The changes returned by the service are arranged in descending order and the WithChanges function that we are using to get the new text for comparison requires the changes to be arranged in ascending order that is why this line has been added.

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs [hunk 9]
```diff
@@ -347,7 +336,7 @@ private static void AssertFormatTargetKind(SyntaxKind kind, string input)
         public static IEnumerable<TextChange> GetTextChanges(SourceText oldText, IEnumerable<LinePositionSpanTextChange> changes)
         {
             var textChanges = new List<TextChange>();
-            foreach( var change in changes)
+            foreach(var change in changes)
```
Comment (rchande) [inline tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: `foreach (var change ...)`

File: tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs (hunk 7 above)
Comment (rchande) [reply tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: Have you tested that? I thought SourceText.WithChanges didn't care about the order.

Comment (rchande) [reply tests/OmniSharp.Roslyn.CSharp.Tests/FormattingFacts.cs]: Never mind: http://source.roslyn.io/#Microsoft.CodeAnalysis/CodeAnalysisResources.Designer.cs,6a0f623f21709018 :)

**Pull request merged**